### ISSUE TITLE
Multiple routes with bearer tokens route to the first defined route

### ISSUE THREAD

**Issue by tailDASHeF** [grafana/grafana#13717]
### What Grafana version are you using? 
Grafana v5.3.1 

### What datasource are you using? 
Prometheus

### What OS are you running grafana on? 
CentOS Linux release 7.5.1804 (Core) Kernel 3.10.0-862.9.1.el7.x86_64

### What did you do? 
Tried to add multiple prometheus routes with bearer tokens like the docs list here http://docs.grafana.org/plugins/developing/auth-for-datasources/#plugin-routes

### What was the expected result? 
That by URL each datasource would be routed correctly

### What happened instead? 
They all routed to the first route in my list, verified by tcpdump

  "routes": [
    {
      "path": "",
      "method": "GET",
      "url": "https://unique-url-1.example.com",
      "headers": [{
        "name": "Authorization", 
        "content": "Bearer redacted" }]
    },
    {
      "path": "",
      "method": "GET",
      "url": "https://unique-url-2.example.com",
      "headers": [{
         "name": "Authorization",
         "content": "Bearer redacted" }]
    },
    {
      "path": "",
      "method": "GET",
      "url": "http://unique-url-2.example.com:9090"
    }],
 
I am trying to have multiple prometheus routes setup, 2 of which (at least) will use a Bearer token - others may use basic auth, certificates, or no auth at all. I need to be able to route these requests to the correct endpoints, I have not found a lot of docs that help with this but what I did find and worked - except for the routing mechanism doesn't see to work as I expected.

**Comment by torkelo** [comment]
they need to have different path values

**Comment by torkelo** [comment]
the routes feature is to allow the data source proxy to route different proxy urls to different services  / apis  , 

so path /v3 is routes to some v3 api service with one set of credentials,  path "v4" is routed to a different api with a second set of credentials.

it is not for multiplexing or causing one request to go to multiple backend and then be merged

**Comment by tailDASHeF** [comment]
Ok that's fine, except my endpoints don't have a path so I have no way to route based upon that. Is there something in the pipeline or any expansion to allow multiple prometheus backends that all require Bearer tokens?

**Comment by torkelo** [comment]
that makes no sense, should it call one at random? With no path there is no way for grafana to know where to route it to. 

> except my endpoints don't have a path 

They dont need to the path is not sent to the proxy it is just a key grafana will use to know which route to use :)

**Comment by tailDASHeF** [comment]
The URL's are different, the path's aren't because prometheus responds just on the root domain. Since the URL is unique, I would assume that it would route based on URL - not path.

for instance https://blah.com not https://blah.com/v1
My urls could be https://asdf.com https://fdsa.com https://qwerty.com. I would assume that it would route the request to the correct datasource because the URL is different for each - not the same.

I'm trying to target multiple prometheus backends with bearer tokens.

In the interim, I have just setup nginx to reverse proxy and add the Authorization: Bearer $TOKEN and have grafana target nginx instead of the real endpoints - this works but is not desired.

**Comment by torkelo** [comment]
I honestly don’t understand what your talking about or trying to do, have you forked the Grafana Prometheus datasource ?

**Comment by tailDASHeF** [comment]
OK, so i'm using grafana with three different datasources that are all type prometheus.

One of my datasources does not require auth, two of them do - and they require Bearer tokens.

According to this doc http://docs.grafana.org/plugins/developing/auth-for-datasources/#plugin-routes the only way to use Bearer tokens is to create routes in the plugins.json file. When you try to create a prometheus datasource you are NOT given an option to specify a Bearer token - you must do so in the plugins.json.

So it was understood to me that routing of the requests was done by URL, which in the stanza I've shared with you is unique by URL, not path. You've explained that routing is done by path, not url - that doesn't work for me or anyone else who would want to use multiple prometheus datasources that require bearer tokens (such as openshift prometheus instances) as they are completely different prometheus servers.

I haven't forked anything or written any code. I've simply reverse proxied my connections to the prometheus instances and handled authentication via nginx because when I attempted to allow grafana to handle authentication via the Bearer token through routes i'm not getting the expected behavior.

So in short, how would you configure 3 unique prometheus datasources two with Bearer authentication , one without. It's clear that routes in plugins.json is not the correct way.

**Comment by torkelo** [comment]
Sorry , you misunderstood those docs, it’s a feature for plugin authors. Your not to modify plugin json for core built in plugins.

We do not support bearer tokens for Prometheus.

**Comment by tailDASHeF** [comment]
Maybe consider this a feature request to support bearer tokens in Prometheus datasources then? It seems others have used this feature that way but likely only had ONE Prometheus datasource.

**Comment by torkelo** [comment]
if you use the datasource provisioning system you can set custom headers  (per data source instance).

```yaml
jsonData:
  httpHeaderName1: "Authorization"
secureJsonData:
  httpHeaderValue1: "Bearer xf5yhfkpsnmgo"
```
